### ISSUE TITLE
Allow define banned-api per file

### ISSUE THREAD

**Issue by Czaki** [astral-sh/ruff#7696]
It will be nice to allow define `banned-api` rules per file pattern. 

In general, there are API that may be useful in tests but should not be used in the library itself:
For example, `warnings.catch_Warnings` because of https://github.com/python/cpython/issues/73858.

On the other hand, I would like to block using `numpy.random` in tests but allow in the library and/or examples (as it leads to random test failure).

If you think it is useful, I could try to contribute (learning rust).

**Comment by MichaReiser** [comment]
That makes sense. Have you tried using a [hierarchical configuration](https://docs.astral.sh/ruff/configuration/#pyprojecttoml-discovery) to achieve your goal? The way this works is that you create a new `ruff.toml` in the test directory that `extend`s your default configuration. The test-specific configuration can override individual settings.

**Comment by Czaki** [comment]
Unfortuantelly in this project (napari), test are not stored in a separate directory but in `_test` modules. And I'm only one of the maintainers. 
(and this simplifies searching for a test of a given function).

**Comment by charliermarsh** [comment]
Really wishing we had ESLint-style file overrides, as in:

```json
{
  "rules": {
    "quotes": ["error", "double"]
  },

  "overrides": [
    {
      "files": ["bin/*.js", "lib/*.js"],
      "excludedFiles": "*.test.js",
      "rules": {
        "quotes": ["error", "single"]
      }
    }
  ]
}
```

**Comment by Czaki** [comment]
cibuildwheel project use toml tables to implement overwrite rules inside pyproject.toml https://github.com/pypa/cibuildwheel/
https://cibuildwheel.readthedocs.io/en/stable/options/#overrides

Maybe it could be an inspiration?

**Comment by MichaReiser** [comment]
> Really wishing we had ESLint-style file overrides, as in:
> 
> ```json
> {
>   "rules": {
>     "quotes": ["error", "double"]
>   },
> 
>   "overrides": [
>     {
>       "files": ["bin/*.js", "lib/*.js"],
>       "excludedFiles": "*.test.js",
>       "rules": {
>         "quotes": ["error", "single"]
>       }
>     }
>   ]
> }
> ```

I've mixed feelings about `overrides`. It is powerful but can become confusing if you mix it with extend. E.g. what happens if you have:


```json
{
   "rules": {
     "quotes": ["error", "double"]
   },
 
   "overrides": [
     {
       "files": ["bin/*.js", "lib/*.js"],
       "excludedFiles": "*.test.js",
       "rules": {
         "quotes": ["error", "single"]
       }
     }
   ]
}
```

and

```json
{
	"extend": "config-above.json",
   "rules": {
     "quotes": ["error", "double"]
   }
}
```

What does now bind stronger, the parent configuration because it's in the `overrides` section or the `extend`? I think we would need to change our configuration more drastically to make "overrides" feel intuitive, similar to what [eslint does with the new flat configuration](https://eslint.org/blog/2022/08/new-config-system-part-2/#glob-based-configs-everywhere)

**Comment by charliermarsh** [comment]
(Making this issue more general -- it would be nice to support something like ESLint's flat configuration, in which you can define configuration that only applies to a subset of files.)

**Comment by DetachHead** [comment]
i'm fine with hierarchical configuration, but the only problem is that i wish it would inherit the `flake8-tidy-imports.banned-api` options from parent directories:
```toml
# ./pyproject.toml
[tool.ruff.flake8-tidy-imports.banned-api]
"foo".msg = "banned project-wide"
```
```toml
# ./tests/ruff.toml
extend = "../pyproject.toml"

[flake8-tidy-imports.banned-api]
"bar".msg = "only banned in tests"
```

```py
# tests/test_foo.py
import foo # no error, because the `banned-api` the `tests` directory doesn't inherit from the global one
```

**Comment by BT-rmartin** [comment]
In my case #7890 we want to configure isort on file-scope basis, and in our particular need the hierarchical solution is not feasible, as we want it for every __init__.py file. Is there any progress on this topic @charliermarsh ?

**Comment by BT-rmartin** [comment]
@charliermarsh Is there any progress on configure isort on file-scope basis?

**Comment by jorenham** [comment]
see https://github.com/astral-sh/ruff/issues/16057 for another unique use-case of this

**Comment by MichaReiser** [comment]
@irdkwmnsb you can already achieve that *if* all legacy files have a common ancestor directory by adding a `ruff.toml` in that ancestor directory.

You need this feature if the files are side by side to non-legacy files.

**Comment by arogozhnikov** [comment]
I have two use-cases for this feature, line-length in notebooks vs plain python files (they are side by side), and an additional 'type' of python file, which also should have a different line-length (also side by side with rest of codebase)
